File: python/python_oop_practice/p_oop_2/chap05/staticex/student.py
class Student:
    serial_num = 1000  # static 변수

    def __init__(self, name = "이름 없음"):
        Student.serial_num += 1  # static 변수 사용시 '클래스명.static변수명' 으로 사용
        self.student_id = Student.serial_num  # 생성자 호출시마다 student_id 자동으로 부여
        self.student_name = name

    def get_serial_num():  # static 메서드
        return Student.serial_num

    def get_student_id(self):
        return self.student_id


def main():
    print(Student.serial_num)  # 객체 생성없이 사용 가능

    student_lee = Student("Lee")
    print(Student.serial_num)  # student_lee.serial_num으로도 접근 가능

    student_kim = Student("Kim")
    print(Student.serial_num)  # student_kim.serial_num으로도 접근 가능

    print(Student.get_serial_num())  # 1002: 현재 부여된 serial_num을 출력 (static 메서드 호출)

    print(student_lee.get_student_id())
    print(student_kim.get_student_id())


if __name__ == '__main__':
    main()

File: java/doit_java/src/chap06/section3/staticex4/Student.java
// 클래스 메서드와 인스턴스 변수
package chap06.section3.staticex4;

public class Student {
    private static int serialNum = 1000;     // private 변수로 변경
    public int studentId;
    public String studentName;
    public int grade;
    public String address;

    public Student() {
        serialNum++;
        studentId = serialNum;
    }

    public String getStudentName() {
        return studentName;
    }

    public void setStudentName(String name) {
        studentName = name;
    }

    public static int getSerialNum() {  // serialNum의 get() 메서드 (클래스 메서드임)
    //  int i = 10;                     // 지역 변수는 가능하지만
    //  studentName = "임채영";          // 인스턴스 변수를 포함할 수 없다. (오류발생)
        return serialNum;
    }

    public static void setSerialNum(int serialNum) {    // serialNum의 set() 메서드
        Student.serialNum = serialNum;
    }
}
